src/linked_list.rs

```diff
@@ -162,59 +167,172 @@ impl<G: AdapterWrapped> List<G> {
     /// # Safety
     ///
     /// Callers must ensure that `existing` points to a valid entry that is on the list.
-    pub unsafe fn insert_after(&mut self, existing: NonNull<G::EntryType>, data: G::Wrapped) {
+    pub unsafe fn insert_after(&mut self, existing: &G::Wrapped, data: G::Wrapped) {
         let ptr = data.into_pointer();
-        unsafe { self.list.insert_after(existing, ptr.as_ref()) }
+        let entry = Wrapper::as_ref(existing);
+        if unsafe { !self.list.insert_after(entry, ptr.as_ref()) } {
+            // If insertion failed, rebuild object so that it can be freed.
+            unsafe { G::Wrapped::from_pointer(ptr) };
+        }
     }
 
     /// Removes the given entry.
     ///
     /// # Safety
     ///
-    /// Callers must ensure that `data` is either on this list. It being on another
+    /// Callers must ensure that `data` is either on this list or in no list. It being on another
     /// list leads to memory unsafety.
     pub unsafe fn remove(&mut self, data: &G::Wrapped) -> Option<G::Wrapped> {
         let entry_ref = Wrapper::as_ref(data);
-        unsafe { self.list.remove(entry_ref) };
-        Some(unsafe { G::Wrapped::from_pointer(NonNull::from(entry_ref)) })
+        if unsafe { self.list.remove(entry_ref) } {
+            Some(unsafe { G::Wrapped::from_pointer(NonNull::from(entry_ref)) })
+        } else {
+            None
+        }
     }
 
     /// Removes the element currently at the front of the list and returns it.
     ///
     /// Returns `None` if the list is empty.
     pub fn pop_front(&mut self) -> Option<G::Wrapped> {
-        let entry_ref = unsafe { self.list.front()?.as_ref() };
-        unsafe { self.list.remove(entry_ref) };
-        Some(unsafe { G::Wrapped::from_pointer(NonNull::from(entry_ref)) })
+        let front = self.list.pop_front()?;
+        // SAFETY: Elements on the list were inserted after a call to `into_pointer `.
+        Some(unsafe { G::Wrapped::from_pointer(front) })
     }
 
-    /// Returns the first element of the list, if one exists.
-    #[inline]
-    pub fn front(&self) -> Option<&G::EntryType> {
-        self.list.front().map(|ptr| unsafe { ptr.as_ref() })
-    }
-
-    /// Returns the last element of the list, if one exists.
-    #[inline]
-    pub fn back(&self) -> Option<&G::EntryType> {
-        self.list.back().map(|ptr| unsafe { ptr.as_ref() })
-    }
-
-    /// Returns a cursor starting on the first (front) element of the list.
-    #[inline]
-    pub fn cursor_front(&self) -> Cursor<'_, G> {
-        self.list.cursor_front()
+    /// Returns a mutable cursor starting on the first (front) element of the list.
+    pub fn cursor_front_mut(&mut self) -> CursorMut<'_, G> {
```

Comment:
Why no cursor_front for read only use?

Reply:
because now we support iterator， no need to supply cursor